### PR TITLE
[MRG+1] Custom solver demo

### PR DESCRIPTION
**Pull request by agramfort**

### CODE REVIEW

File: examples/inverse/plot_custom_inverse_solver.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+This script is educational and shall be used for methods
+evaluations and new developments. It is not meant to be an example
+of good practice to analyse your data.
+"""
+
+from copy import deepcopy
+import numpy as np
+from scipy import linalg
+import mne
+from mne.datasets import sample  # , somato
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: delete comment

File: examples/inverse/plot_custom_inverse_solver.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    It just runs L2 penalized regression and keep the 10 strongest locations
+
+    Parameters
+    ----------
+    M : array, shape (n_channels, n_times)
+        The whitened data.
+    G : array, shape (n_channels, n_dipoles)
+        The gain matrix a.k.a. the forward operator. The number of locations
+        is n_dipoles / n_orient. n_orient will be 1 for a fixed orientation
+        constraint or 3 when using a free orientation model.
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: incomplete docstring

File: examples/inverse/plot_custom_inverse_solver.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    Returns
+    -------
+    X : array, (n_active_dipoles, n_times)
+        The time series of the dipoles in the active set.
+    active_set : array (n_dipoles)
+        Array of bool. Entry j is True if dipole j is in the active set.
+        We have X_full[active_set] == X where X_full is the full X matrix
+        such that `M = G X_full`.
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: double-backticks on code

**Comment by Eric89GXL** [comment]
You didn't add any narrative explaining why two functions are used

**Comment by agramfort** [comment]
@Eric89GXL better?

File: examples/inverse/plot_custom_inverse_solver.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
+# -*- coding: utf-8 -*-
+"""
+================================================
+Source localization with a custom inverse solver
+================================================
+
+The objective of this example is to show how to plug a custom inverse solver
+in MNE in order to facilate empirical comparison with the methods MNE already
+implements (wMNE, dSPM, sLORETA, LCMV, (TF-)MxNE) etc.
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: the `etc.` should be inside the parenthetical, but we can fix this later :)

File: examples/inverse/plot_custom_inverse_solver.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+This script is educational and shall be used for methods
+evaluations and new developments. It is not meant to be an example
+of good practice to analyse your data.
+
+The example makes use of 2 functions ``apply_solver`` and ``solver``
+so changes can be limited to the ``solver`` function in order to
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: Something more like "... to the ``solver`` function (which only takes three parameters: the whitened data, the gain matrix, and the number of orientations)" would help clarify why it's nice

File: examples/inverse/plot_custom_inverse_solver.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    """
+    # Import the necessary private functions
+    from mne.inverse_sparse.mxne_inverse import \
+        (_prepare_gain, _to_fixed_ori, is_fixed_orient,
+         _reapply_source_weighting, _make_sparse_stc)
+
+    all_ch_names = evoked.ch_names
+    # put the forward solution in fixed orientation if it's not already
+    if loose is None and not is_fixed_orient(forward):
+        forward = deepcopy(forward)
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: no need, just do `forward.copy()`

File: examples/inverse/plot_custom_inverse_solver.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    return stc
+
+
+###############################################################################
+# Define your solver
+
+def solver(M, G, n_orient):
+    """Dummy solver
+
+    It just runs L2 penalized regression and keep the 10 strongest locations
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: How hard is it to make this an MCE example? :)

**Comment by agramfort** [comment]
pushed. MCE is for a motivated grad student :)

**Comment by Eric89GXL** [comment]
LGTM +1 for merge

File: examples/inverse/plot_custom_inverse_solver.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        active_set[idx:idx + n_orient] = True
+    X = X[active_set]
+    return X, active_set
+
+###############################################################################
+# Apply your custom solver
+
+# loose, depth = 0.2, 0.8
+loose, depth = 1., 0.
+stc = apply_solver(solver, evoked, forward, noise_cov, loose=loose, depth=depth)
```
Comment (jaeilepp) [inline examples/inverse/plot_custom_inverse_solver.py]: Travis complains about this line being too long.

File: examples/inverse/plot_custom_inverse_solver.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+evoked = mne.read_evokeds(ave_fname, condition=condition, baseline=(None, 0))
+evoked.crop(tmin=0.04, tmax=0.18)
+
+evoked = evoked.pick_types(eeg=False, meg=True)
+# Handling forward solution
+forward = mne.read_forward_solution(fwd_fname, surf_ori=True)
+
+
+###############################################################################
+# Auxilary function to run the solver
```
Comment (Eric89GXL) [inline examples/inverse/plot_custom_inverse_solver.py]: examples/inverse/plot_custom_inverse_solver.py:46: Auxilary  ==> Auxiliary

**Comment by Eric89GXL** [comment]
@jaeilepp can you look and merge if you're happy?

**Comment by jaeilepp** [comment]
Thanks @agramfort

**Pull request merged**